# Samsung Innovation Campus
## Coding and programming
## Tercera prueba de evaluación

Para la realización de este ejercicio se tendrá **2 horas** de tiempo y se podrá sumar hasta **10 puntos**.

**(3 puntos)** Diseña la clase `Coche` que represente coches en venta de un concesionario, determinados por el nombre de la marca, el nombre del modelo, un número de referencia, y un precio base. Estos atributos serán privados. Cuando se calcula el precio de un coche, al precio base se le debe añadir el porcentaje de IVA correspondiente (inicialmente 10 %), un valor que es compartido por todos los objetos de la clase. La clase ofrecerá el siguiente constructor y métodos públicos:
	
~~~ {.py}
▶ def __init__(self,...)
~~~
> Construye un objeto `Coche`. Recibe como parámetros, en el siguiente
> orden, el nombre de la marca, el nombre del modelo, el número de referencia y el precio base del coche.

~~~ {.py}
▶ def marca(self)
▶ def modelo(self)
▶ def referencia(self)
▶ def precio_base(self)
~~~

> Devuelven los valores correspondientes almacenados en el objeto (usa decoradores).

~~~ {.py}
▶ def precio_final(self)
~~~

> Devuelve el precio final del coche, incluyendo el IVA, según la
> siguiente ecuación.
>> $\mathit{PF}=\mathit{PB}+\mathit{PB}\times\mathit{IVA}\div100$

~~~ {.py}
▶ def __str__(self)
~~~

> Devuelve la representación textual del objeto, según el formato del
> siguiente ejemplo:
>> ~~~
>> (123456, Seat Ibiza, 19800€ IVA incluido)
>> ~~~

~~~ {.py}
▶ def __eq__(self,other)
~~~

> Compara el coche actual (self) con el coche recibido como parámetro (other). Dos coches serán iguales si el número de referencia es el mismo. 

In [1]:
# Escribe el código de la clase Coche
class Coche:
    IVA = 10
    
    def __init__(self, marca, modelo, num_referencia, precio_base):
        self.__marca = marca
        self.__modelo = modelo
        self.__referencia = num_referencia
        self.__precio = precio_base

    @property
    def marca(self):
        return self.__marca

    @property
    def modelo(self):
        return self.__modelo
        
    @property
    def referencia(self):
        return self.__referencia
        
    @property
    def precio(self):
        return self.__precio

    def precio_final(self):
        return self.__precio + self.__precio * Coche.IVA / 100
        
    def __str__(self):
        return f"({self.__referencia}, {self.__marca} {self.__modelo}, {self.precio_final():.2f}€ IVA incluido)"


    def __eq__(self, other):
        salida = False
        if isinstance(other, Coche):
            marca_bool = self.__marca == other.marca
            modelo_bool = self.__modelo == other.modelo
            referencia_bool = self.__referencia == other.referencia
            precio_bool = self.__precio == other.precio
            salida = marca_bool and modelo_bool and referencia_bool and precio_bool
        return salida

**(1.5 puntos)** La clase `CocheDeLujo` deriva de la clase `Coche`, por lo que contiene información sobre un determinado coche, pero además, permite especificar el equipamiento extra del coche (p.e. "asientos de cuero"). Será un atributo privado.  La clase ofrecerá el siguiente constructor y métodos públicos:
	
~~~ {.py}
▶ def __init__(self,...)
~~~

> Construye un objeto `CocheDeLujo`. Recibe como parámetros, en el siguiente
> orden, el nombre de la marca, el nombre del modelo, el número de referencia, 
> el precio base del coche y el equipamiento extra.

~~~ {.py}
▶ def equipamiento_extra(self)
~~~

> Devuelve la información del equipamiento extra del coche.

~~~ {.py}
▶ def __str__(self)
~~~

> Devuelve la representación textual del objeto, según el formato del
> siguiente ejemplo:

>> ~~~
>> ((675342, Lamborghini Huracan, 350000€ IVA incluido) asientos de cuero)
>> ~~~



In [2]:
# Escribe el código de la clase CocheDeLujo
class CocheDeLujo(Coche):
    def __init__(self, marca, modelo, num_referencia, precio_base, equipamiento_extra):
        super().__init__(marca, modelo, num_referencia, precio_base)
        self.__equipamiento = equipamiento_extra

    @property
    def equipamiento_extra(self):
        return self.__equipamiento

    def __str__(self):
        return f"({super().__str__()} {self.__equipamiento})"

**(1.25 puntos)** Realiza las siguientes acciones con las clases anteriores.
- Crea tres objetos de la clase `Coche`: 

('Seat', 'Ibiza', 234123, 20000)

('Renault', 'Captur', 123444, 21500)

('Ford', 'Focus', 234123, 22500)

- Crea dos objetos de la clase `CocheDeLujo`:

('Ferrari', 'Testarrosa', 100000, 150000, 'metalizado especial')

('Lamborghini', 'Urus', 200000, 235000, 'asientos de cuero')

- Muestra los objetos por pantalla.

- Compara los tres objetos de la clase `Coche` para ver si son iguales dos a dos.

In [3]:
# Escribe el código para generar objetos Coche y CocheDeLujo

# Creamos objetos de la clase Coche
coche1 = Coche('Seat', 'Ibiza', 234123, 20000)
coche2 = Coche('Renault', 'Captur', 123444, 21500)
coche3 = Coche('Ford', 'Focus', 234123, 22500)

# Creamos objetos de la clase CocheDeLujo
coche_lujo1 = CocheDeLujo('Ferrari', 'Testarrosa', 100000, 150000, 'metalizado especial')
coche_lujo2 = CocheDeLujo('Lamborghini', 'Urus', 200000, 235000, 'asientos de cuero')

# Mostramos los objetos por pantalla
print(coche1)
print(coche2)
print(coche3)
print(coche_lujo1)
print(coche_lujo2)

# Comparamos los objetos de clase Coche dos a dos
print("Comparación de objetos de la clase Coche:")
print("¿Es el coche 1 igual que el coche 2?", coche1 == coche2)
print("¿Es el coche 1 igual que el coche 3?", coche1 == coche3)
print("¿Es el coche 2 igual que el coche 3?", coche2 == coche3)

(234123, Seat Ibiza, 22000.00€ IVA incluido)
(123444, Renault Captur, 23650.00€ IVA incluido)
(234123, Ford Focus, 24750.00€ IVA incluido)
((100000, Ferrari Testarrosa, 165000.00€ IVA incluido) metalizado especial)
((200000, Lamborghini Urus, 258500.00€ IVA incluido) asientos de cuero)
Comparación de objetos de la clase Coche:
¿Es el coche 1 igual que el coche 2? False
¿Es el coche 1 igual que el coche 3? False
¿Es el coche 2 igual que el coche 3? False


**(1 punto)** Obtén la representación textual de los coches anteriores y almacénalos en un fichero con nombre "coches.txt". 

In [4]:
# Escribe el código para almacenar los coches en formato texto en un fichero

representacion_coches = [str(coche1), str(coche2), str(coche3), str(coche_lujo1), str(coche_lujo2)]

# Guardar la representación en el archivo "coches.txt"
with open("coches.txt", "w", encoding = "utf-8") as file:
    for coche in representacion_coches:
        file.write(coche + "\n")

**(2.5 puntos)** Defina la clase `Concesionario`,que represente un concesionario de coches, de tal forma que pueda contener múltiples coches de diversas clases en una lista, todas ellas derivadas de la clase Coche definida en ejercicios anteriores. Este atributo será de ámbito privado.


~~~ {.py}
▶ def __init__(self)
~~~

> Construye un objeto `Concesionario` vacío (sin coches).

~~~ {.py}
▶ def obtener_coches(self)
~~~

> Devuelve la lista de coches del concesionario (usar decorador).


~~~ {.py}
▶ def add_coche(...)
~~~

> Crea un nuevo objeto `Coche` con el nombre de la marca, modelo, número de referencia y 
> precio base recibidos como parámetros. Si el coche está en la lista (si el número de referencia es el mismo) se muestra un mensaje por pantalla, en caso contrario se añade. 

~~~ {.py}
▶ def add_coche_lujo(...)
~~~

> Crea un nuevo objeto `CocheDeLujo` con el nombre de la marca, modelo, número de referencia, 
> precio base y equipamiento extra recibidos como parámetros. Si el coche está en la lista (si el número de referencia es el mismo) se muestra un mensaje por pantalla, en caso contrario se añade. 

~~~ {.py}
▶ def eliminar_coche(...)
~~~

> Dado el número de referencia como parámetro, si el coche está en la lista se elimina, en caso contrario se muestra un mensaje por pantalla. 

~~~ {.py}
▶ def __str__(self)
~~~

> Devuelve la representación textual del objeto, según el formato del
> siguiente ejemplo (sin considerar los saltos de línea):
>> ~~~
>> [(234123, Seat Ibiza, 22000€ IVA incluido)
>>  (123444, Renault Captur, 23650€ IVA incluido)
>>  (234124, Ford Focus, 24750€ IVA incluido)
>>  ((200000, Lamborghini Urus, 258500€ IVA incluido) asientos de cuero) 
>>  ((100000, Ferrari Testarrosa, 16500€ IVA incluido) metalizado especial)] 
>> ~~~


In [5]:
# Escribe el código de la clase Concesionario
class Concesionario:
    def __init__(self):
        self.__coches = [] 

    @property
    def obtener_coches(self):
        return self.__coches
    
    def add_coche(self, marca, modelo, numero_serie, precio):
        coche_nuevo = Coche(marca, modelo, numero_serie, precio)
        existe = False
        for coche in self.__coches:
            if coche.referencia == numero_serie:
                print("El coche ya está en el concesionario.")
                existe = True
        if existe == False:
            self.__coches.append(coche_nuevo)
            print("Coche agregado al concesionario.")
    
    def add_coche_lujo(self, marca, modelo, numero_serie, precio, caracteristicas):
        coche_lujo_nuevo = CocheDeLujo(marca, modelo, numero_serie, precio, caracteristicas)
        existe = False
        for coche in self.__coches:
            if coche.referencia == numero_serie:
                print("El coche ya está en el concesionario.")
                existe = True
        if existe == False:
            self.__coches.append(coche_lujo_nuevo)
            print("Coche de lujo agregado al concesionario.")
    
    def eliminar_coche(self, numero_serie):
        coche_a_eliminar = None
        for coche in self.__coches:
            if coche.referencia == numero_serie:
                coche_a_eliminar = coche
        if coche_a_eliminar:
            self.__coches.remove(coche_a_eliminar)
            print("Coche eliminado correctamente.")
        else:
            print("El coche de referencia", numero_serie, "no está en el concesionario.")
    
    def __str__(self):
        coches_str = "["
        for coche in self.__coches:
            coches_str += str(coche) + ",\n"
        coches_str = coches_str[:-2] + "]"
        return coches_str

In [6]:
# Pruebas de Concesionario

concesionario = Concesionario()

concesionario.add_coche('Seat', 'Ibiza', 234123, 20000)
concesionario.add_coche('Renault', 'Captur', 123444, 21500)
concesionario.add_coche('Ford', 'Focus', 234124, 22500)

concesionario.add_coche_lujo('Lamborghini', 'Urus', 200000, 235000, 'asientos de cuero')
concesionario.add_coche_lujo('Ferrari', 'Testarrosa', 100000, 150000, 'metalizado especial')

print(concesionario)

concesionario.eliminar_coche(123444)
concesionario.eliminar_coche(234125)  # Un número de serie inexistente

print(concesionario)

Coche agregado al concesionario.
Coche agregado al concesionario.
Coche agregado al concesionario.
Coche de lujo agregado al concesionario.
Coche de lujo agregado al concesionario.
[(234123, Seat Ibiza, 22000.00€ IVA incluido),
(123444, Renault Captur, 23650.00€ IVA incluido),
(234124, Ford Focus, 24750.00€ IVA incluido),
((200000, Lamborghini Urus, 258500.00€ IVA incluido) asientos de cuero),
((100000, Ferrari Testarrosa, 165000.00€ IVA incluido) metalizado especial)]
Coche eliminado correctamente.
El coche de referencia 234125 no está en el concesionario.
[(234123, Seat Ibiza, 22000.00€ IVA incluido),
(234124, Ford Focus, 24750.00€ IVA incluido),
((200000, Lamborghini Urus, 258500.00€ IVA incluido) asientos de cuero),
((100000, Ferrari Testarrosa, 165000.00€ IVA incluido) metalizado especial)]


**(0.75 puntos)** Obtén la lista de coches del concesionario y almacénalo en un fichero binario usando la librería *pickle*. Cárgalo de nuevo en otra variable y muéstralo por pantalla.

In [7]:
# Escribe el código para obtener la lista y guardar en formato binario
import pickle

lista_coches = concesionario

f = open('coches_concesionario.pkl','wb')
pickle.dump(lista_coches, f)
f.close()
    
nueva_lista_coches = pickle.load(open('coches_concesionario.pkl','rb'))
print(nueva_lista_coches)

[(234123, Seat Ibiza, 22000.00€ IVA incluido),
(234124, Ford Focus, 24750.00€ IVA incluido),
((200000, Lamborghini Urus, 258500.00€ IVA incluido) asientos de cuero),
((100000, Ferrari Testarrosa, 165000.00€ IVA incluido) metalizado especial)]
